pkg/storaged/nfs-mounts.py

```diff
@@ -0,0 +1,344 @@
+
+import select
+import re
+import sys
+import json
+
+import os
+import ctypes
+import struct
+import subprocess
+import traceback
+
+# We monitor all NFS remotes listed in /etc/fstab and in /etc/mtab.
+# If a entry from mtab is also found in fstab, we report only the
+# fstab entry and mark it is "mounted".
+
+IN_CLOSE_WRITE = 0x00000008
+IN_IGNORED     = 0x00008000
+
+class Inotify:
+    def __init__(self):
+        self._libc = ctypes.CDLL(None, use_errno=True)
+        self._get_errno_func = ctypes.get_errno
+
+        self._libc.inotify_init.argtypes = []
+        self._libc.inotify_init.restype = ctypes.c_int
+        self._libc.inotify_add_watch.argtypes = [ctypes.c_int, ctypes.c_char_p,
+                                                 ctypes.c_uint32]
+        self._libc.inotify_add_watch.restype = ctypes.c_int
+        self._libc.inotify_rm_watch.argtypes = [ctypes.c_int, ctypes.c_int]
+        self._libc.inotify_rm_watch.restype = ctypes.c_int
+
+        self.fd = self._libc.inotify_init()
+
+    def add_watch(self, path, mask):
+        path = ctypes.create_string_buffer(path.encode(sys.getfilesystemencoding()))
+        wd = self._libc.inotify_add_watch(self.fd, path, mask)
+        if wd < 0:
+            sys.stderr.write("can't add watch for %s: %s\n" % (path, os.strerror(self._get_errno_func())))
+        return wd
+
+    def rem_watch(self, wd):
+        if self._libc.inotify_rm_watch(self.fd, wd) < 0:
+            sys.stderr.write("can't remove watch: %s\n" % (os.strerror(self._get_errno_func())))
+
+    def process(self, callback):
+        buf = os.read(self.fd, 4096)
+        pos = 0
+        while pos < len(buf):
+            (wd, mask, cookie, name_len) = struct.unpack('iIII', buf[pos:pos+16])
+            pos += 16
+            (name,) = struct.unpack('%ds' % name_len, buf[pos:pos + name_len])
+            pos += name_len
+            callback(wd, mask, name.decode().rstrip('\0'))
+
+class Watcher:
+    def __init__(self, path):
+        self.inotify = Inotify()
+        self.path = path
+        self.setup()
+
+    def setup(self):
+        self.wd = self.inotify.add_watch(self.path, IN_CLOSE_WRITE)
+
+    def process(self, callback = None):
+        def event(wd, mask, name):
+            if callback:
+                callback()
+            if mask & IN_IGNORED:
+                self.setup()
+        self.inotify.process(event)
+
+def field_escape(str):
+    return str.replace("\\", "\\134").replace(" ", "\\040").replace("\t", "\\011")
+
+def field_unescape(str):
+    return re.sub("\\\\([0-7]{1,3})", lambda m: chr(int(m.group(1), 8)), str)
+
+def parse_tab(name):
+    entries = [ ]
+    for line in open(name, "r").read().split("\n"):
+        sline = line.strip()
+        if sline == "" or sline[0] == "#":
+            continue
+        fields = list(map(field_unescape, re.split("[ \t]+", sline)))
+        if len(fields) > 0 and ":" in fields[0]:
+            entries.append(fields)
+    return entries
```

Comment:
As you are already using ctypes, would it be possible/easier to use `{get,set,add}mntent()`? The `FILE*` pointers can just be passed around as opaque values, so they shouldn't hurt much, and `struct mntent` is just `char*` and `int` and thus should map to `ctypes` fairly well?

Reply:
Good point.

Reply:
Hmm, there doesn't seem to be a way to update an entry in place, or remove one.  In fact, udisks2 also does its own parsing and unparsing of /etc/fstab.